### PR TITLE
Use Bank Account Checker API to compare sort codes / account numbers

### PR DESCRIPTION
**Pull request by mattandrews**
This means a valid bank account is required to use the form. A pair of non-matching sort code / account numbers will fail validation.

On the other hand, if the API call fails for any other reason (eg. downtime, token expiry, etc) we treat it as a success and proceed.

Generally this adds the concept of a `preflightCheck`, which is an async function on a `step` which returns a `Promise`. If the function throws an error (eg. returns an array of error objects) then these are rendered back to the user:

image

https://trello.com/c/wAiZeexS/534-use-bank-account-checker-to-validate-bacs-is-allowed

### CODE REVIEW

**Comment by mattandrews** [comment]
For testing, here are some (test) account details that are valid:

```
{
    sortCode: 308087,
    bankAccount: 25337846
}
```

File: common/bank-api.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+'use strict';
+const request = require('request-promise-native');
+const { BANK_API } = require('../common/secrets');
+
+const checkBankAccountDetails = (sortCode, accountNumber) => {
```
Comment (davidrapson) [inline common/bank-api.js]: As this is on a singe function I'd probably export it directly

```js
module.exports = function checkBankAccountDetails(sortCode, accountNumber) { … }
```

File: controllers/apply/awards-for-all/form.js [hunk 2]
```diff
@@ -620,7 +623,76 @@ module.exports = function({ locale, data = {} }) {
...
+                preflightCheck: function() {
+                    const sortCode = get('bankSortCode')(data);
+                    const accountNumber = get('bankAccountNumber')(data);
+
+                    return new Promise((resolve, reject) => {
+                        checkBankAccountDetails(sortCode, accountNumber)
+                            .then(status => {
+                                if (status.code === 'UNKNOWN') {
+                                    // If this API does anything weird, assume all is well
+                                    logger.info(
```
Comment (davidrapson) [inline controllers/apply/awards-for-all/form.js]: Mind adding a tag for the logger like in https://github.com/biglotteryfund/blf-alpha/pull/2020 ?

File: controllers/apply/awards-for-all/form.js [hunk 3]
```diff
@@ -620,7 +623,76 @@ module.exports = function({ locale, data = {} }) {
...
+                                    ]);
+                                }
+                                // Otherwise everything is all good
+                                return resolve();
+                            })
+                            .catch(err => {
+                                logger.info(
+                                    'User bank details check: API call failed',
+                                    {
+                                        error: err
```
Comment (davidrapson) [inline controllers/apply/awards-for-all/form.js]: Would it be better to capture the error part in Sentry?

File: controllers/apply/awards-for-all/form.js [hunk 4]
```diff
@@ -620,7 +623,76 @@ module.exports = function({ locale, data = {} }) {
...
                         ]
                     }
-                ]
+                ],
+                preflightCheck: function() {
+                    const sortCode = get('bankSortCode')(data);
+                    const accountNumber = get('bankAccountNumber')(data);
+
+                    return new Promise((resolve, reject) => {
+                        checkBankAccountDetails(sortCode, accountNumber)
```
Comment (davidrapson) [inline controllers/apply/awards-for-all/form.js]: Could make this async/await

File: controllers/apply/awards-for-all/form.js [hunk 5]
```diff
@@ -624,12 +629,12 @@ module.exports = function({ locale, data = {} }) {
...
                 ],
-                preflightCheck: function() {
+                preflightCheck: async function() {
                     const sortCode = get('bankSortCode')(data);
                     const accountNumber = get('bankAccountNumber')(data);
 
-                    return new Promise((resolve, reject) => {
-                        checkBankAccountDetails(sortCode, accountNumber)
+                    return new Promise(async (resolve, reject) => {
+                        await checkBankAccountDetails(sortCode, accountNumber)
```
Comment (davidrapson) [inline controllers/apply/awards-for-all/form.js]: This syntax isn't correct, needs to be a `try/catch` if using async await.

**Pull request merged**